Rust port of mediasoup!

nazar-pc opened the pull request:
What this massive PR includes:
* Complete Rust port of the TypeScript mediasoup library
* Port of the relevant tests from TypeScript
* Some Rust-specific tests
* Built-in documentation (as generally expected in Rust ecosystem)
* Echo example app (server in Rust + client in TypeScript)
* CI config that checks formatting with rustfmt, lints code with clippy and runs tests

There are minor cleanups in worker and existing tests, but nothing significant (let me know if that is an issue).
They are primarily related to stricter typing generally used in Rust.

Since there is #397 open and adding CI for Rust into existing setup would be challenging, I decided to use GitHub actions instead, hopefully that is OK.

Code style used is default for Rust ecosystem with no tweaks. For TypeScript code used in echo example (the only one so far) I used already established code style from this repository to make it consistent.

I acknowledge how big this is and would be happy to address any feedback there might be. Please start with documentation on docs.rs, it explains some high-level information about this port.

References:
* discussion on forum: https://mediasoup.discourse.group/t/mediasoup-library-in-rust/1803
* releases on crates.io: https://crates.io/crates/mediasoup
* documentation on docs.rs: https://docs.rs/mediasoup

As discussed with Iñaki, there will be some complementary activities related to this around website documentation and such.

nazar-pc committed:
Create channel struct in its own module to split things up a bit

nazar-pc committed:
Move log level and tags structs into separate module for reusability without awkward imports

nazar-pc committed:
Draft some of the messages sent to the worker as data structures, but far from all

nazar-pc committed:
Channel request prototyping

nazar-pc committed:
Refactor work with channels to something nicer

nazar-pc committed:
Make sure internal channels are closed when channel struct is dropped

nazar-pc committed:
Control message length in channel

nazar-pc committed:
Derive `Clone` on channel

nazar-pc committed:
Timeout on requests

nazar-pc committed:
Tiny cleanup

nazar-pc committed:
Request method now works

nazar-pc committed:
First working request method on worker: dumping

nazar-pc committed:
Smarter request/response construction where request is bound to response, more careful visibility settings

nazar-pc committed:
Generate request/response data structures with macro

nazar-pc committed:
Implement resource usage request

nazar-pc committed:
Tiny refactoring

nazar-pc committed:
Tiny macro refactoring

nazar-pc committed:
Another round of moving things around, method for updating worker settings

nazar-pc committed:
Router creation API with some ORTC-related data structures, stub for Router itself

nazar-pc committed:
Refactor macro for request/response handling one more time

nazar-pc committed:
Make sure to send sigterm and not kill process when worker is dropped

nazar-pc committed:
Add methods to subscribe to new router and closed events on worker

nazar-pc committed:
Add app data to worker and router

nazar-pc committed:
Tiny tweaks

nazar-pc committed:
New type for router ID

nazar-pc committed:
RouterId moved to data structures, Drop impl for Router extended with closing router on worker side, some trivial methods on Router, methods comments

nazar-pc committed:
Method for attaching event handler for new workers on manager

nazar-pc committed:
Router dumping implementation

nazar-pc committed:
Start prototyping WebRTC transport

nazar-pc committed:
In Rust version we don't need to keep track internally, so derive Clone on Worker and Router instead

nazar-pc committed:
Refactor modules and data structure locations, tweak documentation a bit in WorkerManager struct

nazar-pc committed:
Tiny simplification

nazar-pc committed:
Update dependencies

nazar-pc committed:
Make sure parent data structure is not dropped when creating worker, router or transport is created

nazar-pc committed:
Add `Worker::connect_died()` method

nazar-pc committed:
Add TODO

nazar-pc committed:
Ability to dump transport and get its stats

nazar-pc committed:
Tiny renaming

nazar-pc committed:
Add comment about how long worker manager, worker and router will be kept alive

nazar-pc committed:
Improved generics syntax for messages generation

nazar-pc committed:
Implementation for more transport methods

nazar-pc committed:
Move implementation of generic transport requests from WebRTC transport into trait that can be implemented on any transport

nazar-pc committed:
Start prototyping producer creation

nazar-pc committed:
Extend list of derived impls for many types, make properties that need to be public, well, public, skip optional fields from serialization as it is done in TypeScript version

nazar-pc committed:
Add an ability to create subscriptions on internal channel, will be useful later

nazar-pc committed:
ICE restart on WebRTC transport (not tested), minor tweaks

nazar-pc committed:
Refactor channel subscriptions, prepare for receiving notifications in WebRTC transport

nazar-pc committed:
Theoretically working notification handling in WebRTC transport

nazar-pc committed:
Minor refactoring to use `async_trait` in `TransportImpl`, start prototyping `TransportImpl::produce_impl()`

nazar-pc committed:
Better typing for RTP parameters, RTP parameters validation in `TransportImpl::produce_impl()`

nazar-pc committed:
Make RTP parameters module public

nazar-pc committed:
More prototyping in `TransportImpl::produce_impl()`

nazar-pc committed:
More prototyping in `TransportImpl::produce_impl()`

nazar-pc committed:
More prototyping in `TransportImpl::produce_impl()`

nazar-pc committed:
Theoretically working `TransportImpl::produce_impl()` implementation

nazar-pc committed:
Actually working `TransportImpl::produce_impl()` implementation, fixes to router creation

nazar-pc committed:
Add finalized versions of RTP codec capabilities and RTP capabilities such that preferred payload type is not both optional and non-optional depending on context

nazar-pc committed:
Cleanup some TODOs

nazar-pc committed:
Implement `Drop` on WebRTC transport that closes transport on worker side

nazar-pc committed:
Add some of the previously missing checks to `ortc.rs`

nazar-pc committed:
Stubs for a bunch of producer methods

nazar-pc committed:
Theoretically working notifications handling in Producer

nazar-pc committed:
Producer dump implementation, minor tweaks

nazar-pc committed:
Theoretically working producer stats implementation

nazar-pc committed:
Implement producer pausing/resuming

nazar-pc committed:
Implement enabling trace event types on producer and transport

nazar-pc committed:
Start prototyping of consumer creation

nazar-pc committed:
More prototyping of consumer creation

nazar-pc committed:
Working consumer creation, beginning of Consumer struct prototyping, minor tweaks

nazar-pc committed:
Continue Consumer struct prototyping, minor tweaks

nazar-pc committed:
Dumping implementation for Consumer, clarifications for transport and producer dumping

nazar-pc committed:
Stats implementation for Consumer

nazar-pc committed:
Consumer pause/resume, minor fixes

nazar-pc committed:
Implement Consumer set preferred layers

nazar-pc committed:
Implement Consumer set/unset priority

nazar-pc committed:
Reduce code duplication slightly

nazar-pc committed:
Implement Consumer request key frame

nazar-pc committed:
Update dependencies, switch from `rand` crate to `fastrand` to reduce dependency tree

nazar-pc committed:
Handle notifications and enabling trace events, tiny tweaks

nazar-pc committed:
Make modules look nice in docs

nazar-pc committed:
Basic docs about crate in `lib.rs`, move things out of `data_structures.rs` and tweak visibility

nazar-pc committed:
Rough sketch of DataProducer struct and surrounding code

nazar-pc committed:
Prototyping data producer creation on transport

nazar-pc committed:
Data producer creation working (data producer itself is not), minor fixes

nazar-pc committed:
Data producer done (without direct transport support), minor tweaks

nazar-pc committed:
Rough sketch of DataConsumer struct and surrounding code

nazar-pc committed:
Prototyping data consumer creation on transport

nazar-pc committed:
Data consumer creation working (data consumer itself is not)

nazar-pc committed:
Data consumer done (without direct transport support and not send methods whose purpose is unclear)

nazar-pc committed:
Update dependencies

nazar-pc committed:
Refactor producer/data producer tracking logic from WebRTC transport creation method

nazar-pc committed:
Fix compilation, minor renaming

nazar-pc committed:
Fix compilation

nazar-pc committed:
Update dependency

nazar-pc committed:
Plain transport implementation

nazar-pc committed:
Bump version

nazar-pc committed:
Update Cargo.toml to include link to docs and repository

nazar-pc committed:
Reduce code duplication by moving more logic into TransportImpl trait

nazar-pc committed:
Refactor `connect_` methods to `on_`

nazar-pc committed:
Improve event handling ergonomics by having `HandlerId` returned when attaching a handler, which when dropped will remove handler automatically, tiny tweaks

nazar-pc committed:
Bump version

nazar-pc committed:
Remove `event_handlers` module that is now published as standalone crate

nazar-pc committed:
Add `Router::can_consume()` method

nazar-pc committed:
Prototyping audio level observer creation on router

nazar-pc committed:
Finish audio level observer

nazar-pc committed:
Create separate file for smoke tests

nazar-pc committed:
Remove unnecessary code

nazar-pc committed:
Bump version

nazar-pc committed:
More consistent names for event handlers

nazar-pc committed:
Payload channel prototype

nazar-pc committed:
Refactor `Mutex<bool>` to `AtomicBool` instead

nazar-pc committed:
Distinction between producer and data producer types (Regular/Direct) and `send` method for Producer and DataProducer created on Direct transport

nazar-pc committed:
Direct transport implementation, Direct transport-related features on consumer/producer/data consumer/data producer

nazar-pc committed:
Implement missing piece of code in producer creation step

nazar-pc committed:
Pipe transport implementation (not tested)

nazar-pc committed:
Implement `Router.pipe_producer_to_router()` and `Router.pipe_data_producer_to_router()`, various tweaks

nazar-pc committed:
Fix order of destruction

nazar-pc committed:
Cleanup

nazar-pc committed:
Bump version

nazar-pc committed:
Typo

nazar-pc committed:
Apply or suppress clippy suggestions, minor tweaks

nazar-pc committed:
Switch to Mutex implementation from `parking_lot` crate

nazar-pc committed:
Minor optimizations

nazar-pc committed:
Update dependencies

nazar-pc committed:
Remove some `unsafe` usage that is not critically necessary

nazar-pc committed:
Switch from `String` to `IpAddr` for listening IP

nazar-pc committed:
Switch more IP addresses from `String` to `IpAddr`

nazar-pc committed:
Remove fixed TODOs

nazar-pc committed:
Print dumps into stderr instead of stdout

nazar-pc committed:
Better invariants protection for various options

nazar-pc committed:
Create an enum for RTP header extension URIs

nazar-pc committed:
Add unsupported RTP header extension URI variant (useful for de-serialization)

nazar-pc committed:
Refactor `RtcpFeedback` into en enum with custom serialization/deserialization for convenience

nazar-pc committed:
Fix parsing scalability modes, add corresponding tests

nazar-pc committed:
Remove some irrelevant TODOs

nazar-pc committed:
Remove some irrelevant TODOs, switch from `String` to `IpAddr` in a few more places

nazar-pc committed:
Remove some irrelevant TODOs, improve typing of worker settings, some more useful derives

nazar-pc committed:
Minor refactoring

nazar-pc committed:
AudioLevelObserver tests ported from TypeScript, minor tweaks to types

nazar-pc committed:
One consumer test, minor fixes and types improvements

nazar-pc committed:
Derive a bunch of useful implementations on data structures fix some visibilities

nazar-pc committed:
Remaining consumer tests, type tweaks, minor tweak to mediasoup itself to match types properly

nazar-pc committed:
First batch of producer tests, some tweaks and fixes

nazar-pc committed:
Remaining producer tests and minor tweaks

nazar-pc committed:
Bump version

nazar-pc committed:
Fix handling for `bufferedamountlow` notification in `DataConsumer`, add `on_data_producer_close()` method on `DataConsumer`

nazar-pc committed:
Remove redundant log output in tests unless log level is specified explicitly

nazar-pc committed:
Refactor `RtpCodecParametersParameters` from `BTreeMap` for improved ergonomics

nazar-pc committed:
Improve tests readability

nazar-pc committed:
Add useful derives to error enums

nazar-pc committed:
Integrate ported dependency `h264-profile-level-id` and unlock corresponding Producer test case

nazar-pc committed:
Port ortc tests, minor refactoring

nazar-pc committed:
Port worker tests, minor refactoring

nazar-pc committed:
Port router tests, minor refactoring, some new TODOs

nazar-pc committed:
Fix memory leak

nazar-pc committed:
`AudioLevelObserver::on_router_close()` method and corresponding tests

nazar-pc committed:
`::on_router_close()` method on transports

nazar-pc committed:
`::on_router_close()` method on producer, consumer, data_producer and data_consumer alongside corresponding tests

nazar-pc committed:
Fix cargo doc warnings

nazar-pc committed:
Apply clippy suggestions

nazar-pc committed:
Update dependencies, bump version

nazar-pc committed:
Refactor scalability modes to idiomatic Rust

nazar-pc committed:
Initial WebRTC transport test, minor tweaks

nazar-pc committed:
Port 2 more WebRTC transport tests, minor tweaks

nazar-pc committed:
Port 4 more WebRTC transport tests, minor tweaks and fix to connection

nazar-pc committed:
Port remaining WebRTC transport tests

nazar-pc committed:
Plain transport tests, minor tweaks and fixes

nazar-pc committed:
Pipe transport tests, minor tweaks and fixes (2 tests are not passing and are commented-out)

nazar-pc committed:
Bump version

nazar-pc committed:
Direct transport tests, minor tweaks and fixes

nazar-pc committed:
Fix and unlock 2 Pipe transport tests, expose `closed()` method on consumer/data consumer/producer/data producer

nazar-pc committed:
Add `closed()` method on transports, handle router closing better

nazar-pc committed:
Add `closed()` method on router and worker

nazar-pc committed:
Include `closed()` check in tests

nazar-pc committed:
Remove fields from C++ that are not used for anything other than tests and even there they were present conflicting with expected type

nazar-pc committed:
Data consumer tests, minor fixes

nazar-pc committed:
Data producer tests, minor tweaks

nazar-pc committed:
Remove unnecessary code

nazar-pc committed:
Bump version

nazar-pc committed:
Fix racy deadlock caused by drop order in event handlers (fixed with `event-listener-primitives` update)

nazar-pc committed:
Fix circular references in `Router::pipe_*_to_router()` implementation, minor tweaks

nazar-pc committed:
Bump version

nazar-pc committed:
Better error message when worker process dies

nazar-pc committed:
Avoid spawning multiple workers concurrently, this results in racy issues

nazar-pc committed:
Fix racy test

nazar-pc committed:
Better heavier typing for DtlsFingerprint

nazar-pc committed:
Main documentation page update, bump version

nazar-pc committed:
Switch to parking_lot for Mutex implementation in worker spawning function

nazar-pc committed:
Move `TransportGeneric::enable_trace_event()` to `Transport::enable_trace_event()`, documentation updates for a bunch of things

nazar-pc committed:
Use relaxed atomic operation ordering for MID generation, strongly consistent is not needed here

nazar-pc committed:
More documentation updates, minor tweaks

nazar-pc committed:
More documentation updates, minor tweaks

nazar-pc committed:
More documentation updates, minor tweaks

nazar-pc committed:
More documentation updates, minor tweaks

nazar-pc committed:
More documentation updates, minor tweaks

nazar-pc committed:
More documentation updates, minor tweaks

nazar-pc committed:
More documentation updates, minor tweaks

nazar-pc committed:
Honor original casing of project name

nazar-pc committed:
Readme and lib.rs update

nazar-pc committed:
Fix links, update readme with badges

nazar-pc committed:
Add info about worker binary to docs

nazar-pc committed:
Bump version

nazar-pc committed:
Update readme link and bump version

nazar-pc committed:
Add public weak versions of consumer/producer/data consumer/data consumer, audio level observer and transports, tiny tweaks

nazar-pc committed:
Tests for weak versions

nazar-pc committed:
Improve usability of piping producers and data producers to other routers

nazar-pc committed:
Apply clippy suggestions

nazar-pc committed:
Tiny documentation improvement

nazar-pc committed:
Bump version

nazar-pc committed:
Update dependencies

nazar-pc committed:
Fix a few typos

nazar-pc committed:
Echo example app

nazar-pc committed:
Simplify generic code, replace generic parameters on traits with associated types since we only want one implementation of those

nazar-pc committed:
Optimize allocations and memory copying in Channel/PayloadChannel when sending to worker, minimize copying

nazar-pc committed:
Bump version

nazar-pc committed:
Implement `Clone` on `Box<dyn Transport>` that may be convenient in some cases

nazar-pc committed:
Move common transport `on_*` methods to `Transport` trait for convenience and make them accept boxed functions as arguments

nazar-pc committed:
Minor refactoring

nazar-pc committed:
Make common rtp observer `on_*` methods accept boxed functions as arguments

nazar-pc committed:
Enum for video rotation

nazar-pc committed:
Add categories, bump version

nazar-pc committed:
Add equality derives for rotation enum and fix its visibility

nazar-pc committed:
Remove most of async mutex usages where it is not needed for simpler API and better performance

nazar-pc committed:
Optimize allocations in notifications-related code

nazar-pc committed:
Fix racy deadlock

nazar-pc committed:
Update dependency, bump version

nazar-pc committed:
Add pipe boolean option support

nazar-pc committed:
Ignore errors happening when worker is in the process of being destroyed anyway

nazar-pc committed:
Implement buffering for notifications for entities that are being created so we don't miss notifications due to race conditions

nazar-pc committed:
Call the callback provided in `on_close` right away if the instance is already closed, since otherwise it will never be fired anyway

nazar-pc committed:
Bump version

nazar-pc committed:
Extract inline modules with unit tests for compilation improvements

nazar-pc committed:
Refactor integration tests such that compilation is faster and they run quicker with more parallelism

nazar-pc committed:
Fix formatting

nazar-pc committed:
GitHub Actions workflow for Rust

nazar-pc committed:
Create workspace, fix/suppress clippy suggestions

nazar-pc committed:
Fix or suppress nightly clippy warnings

nazar-pc committed:
Fix tests after rebase against upstream

nazar-pc committed:
Tweak GitHub Actions workflow to hash Cargo.toml instead

nazar-pc committed:
Change links to upstream versatica repository

ibc commented on .gitignore:
```diff
@@ -37,3 +37,10 @@
 # Vistual Studio Code stuff.
 /.vscode/
+
+# JetBrains IDE stuff
```
```suggestion
# JetBrains IDE stuff.
```

ibc commented on test/test-Consumer.js:
```diff
@@ -681,10 +681,7 @@ test('consumer.dump() succeeds', async () =>
 				{
 					ssrc : videoConsumer.rtpParameters.encodings[0].rtx.ssrc
 				},
-				scalabilityMode : 'S4T1',
-				spatialLayers   : 4,
-				temporalLayers  : 1,
-				ksvc            : false
```
Why are these fields removed? Actually they represent real fields that must be set.

nazar-pc replied:
The reason is that these are inconsistent with defined types.
https://mediasoup.org/documentation/v3/mediasoup/rtp-parameters-and-capabilities/#RtpParameters doesn't have `spatialLayers`, `temporalLayers` and `ksvc` in it and thus identical test in Rust wasn't possible to write because it just wouldn't compile.

Since this is a dump and the only place where I found these properties are used were tests, I decided to remove them from tests and from corresponding place in the worker itself. At the end of the day `scalabilityMode` already encapsulates all three and you can parse it into those fields if really needed.

ibc replied:
It's not a dump, these fields are part of "our" `RtpEncodingParameters` type. We may have to properly document them indeed, but they are needed. In fact we use them in production.

BTW `scalabilityMode` is a real W3 field defined in an additional spec: https://www.w3.org/TR/webrtc-svc/

nazar-pc replied:
Hm... this particular `data` is a dump as you can see from the call above and `data.rtpParameters` indeed comes from `RtpEncodingParameters::FillJson` in the worker, which is an implementation for dumping said object.

Neither documentation nor TypeScript code has `spatialLayers`, `temporalLayers` or `ksvc` here.

I did NOT remove `scalabilityMode` (which I think is the source of confusion here). It is red on this side of the diff, but it is also green on the other side. Is that your concern here?

ibc commented on test/test-WebRtcTransport.js:
```diff
@@ -96,15 +96,12 @@ test('router.createWebRtcTransport() succeeds', async () =>
-			isDataChannel      : true,
-			sctpBufferedAmount : 0,
-			sendBufferSize     : 262144
+			port           : 5000,
+			OS             : 2048,
+			MIS            : 2048,
+			maxMessageSize : 1000000
```
Same here, why so many fields removed?

nazar-pc replied:
This property represents https://mediasoup.org/documentation/v3/mediasoup/sctp-parameters/#SctpParameters struct, which doesn't have fields `isDataChannel`, `sctpBufferedAmount` and `sendBufferSize`. Identical Rust version of the test again didn't compile due to type mismatch.

These were again only used in tests and never described in public API, so I decided to clean that up and remove both from tests and worker itself for consistency.

If those properties are important to check, they need to be extracted from dump or elsewhere, not placed into `SctpParameters` where they don't belong.

ibc reviewed (commented):
Amazing. I must do a proper review and it will take some days before I finish. For now just a few comments:

- I'd like to move current Travis-CI to GitHub CI entirely and not just for Rust.
- I don't like the idea of including frontend examples in `mediasoup/rust/` folder.
- I don't understand why there is `rust/.gitignore` instead of just adding entries to main `.gitignore`.
- Why is mediasoup rust version 0.5.0?

jmillan commented on worker/src/RTC/SctpAssociation.cpp:
```diff
@@ -356,15 +356,6 @@ namespace RTC
 
 		// Add maxMessageSize.
 		jsonObject["maxMessageSize"] = this->maxSctpMessageSize;
-
```
Why this removals?

nazar-pc replied:
Answered in reply to corresponding test change above.

nazar-pc committed:
.gitignore tweaks

nazar-pc commented:
> I'd like to move current Travis-CI to GitHub CI entirely and not just for Rust.

Files for Rust and Node.js will most likely be different, so one doesn't depend on the other. I will try to find time for Node.js workflow in a separate PR. Here I just didn't want to submit it without CI in place.

> I don't like the idea of including frontend examples in `mediasoup/rust/` folder.
> I don't understand why there is `rust/.gitignore` instead of just adding entries to main `.gitignore`.

Moved `.gitignore` entries to the root in last commit

nazar-pc commented:
> I'd like to move current Travis-CI to GitHub CI entirely and not just for Rust.

Files for Rust and Node.js will most likely be different, so one doesn't depend on the other. I will try to find time for Node.js workflow in a separate PR. Here I just didn't want to submit Rust version without CI in place from the beginning.

> I don't like the idea of including frontend examples in `mediasoup/rust/` folder.

There is a convension in Rust (supported by Cargo) to have `examples` folder, so you can do `cargo run --example echo`.
So it felt natural to have frontend directory nearby.

There is no identical backend example in TypeScript so it seems fine to me to have it there.

I have no stong feelings about it if you want me to move it somewhere.
I can even move Rust examples and corresponding frontend to the root or maybe even to separate repo if you want (but that would be harder to maintain, cargo workspaces are great and everything is compiled together in CI).

> I don't understand why there is `rust/.gitignore` instead of just adding entries to main `.gitignore`.

Moved `.gitignore` entries to the root in last commit

> Why is mediasoup rust version 0.5.0?

This is less similar to TypeScript library and more similar to mediasoup-client case.
Rust version will have different things that will be considered a breaking change (for instance #503 broke dump in Rust because they are heavily typed there) and I don't think versioning will line-up perfectly.
And there will probably be more things that will break until it reaches the level of API stability of mediasoup itself.

I am open to suggestions how to make it clearer which version of the worker/mediasoup Rust version corresponds to (I'm also exploring how to turn worker into Rust crate to it can be installed as a dependency automatically).

Updated ticket with similarities/differences with version info.

ibc commented:
> > I'd like to move current Travis-CI to GitHub CI entirely and not just for Rust.
> 
> Files for Rust and Node.js will most likely be different, so one doesn't depend on the other. I will try to find time for Node.js workflow in a separate PR. Here I just didn't want to submit Rust version without CI in place from the beginning.

Thanks. In fact we are having problems with Travis-CI due to its new limits.
 
> > I don't like the idea of including frontend examples in `mediasoup/rust/` folder.
> 
> There is a convension in Rust (supported by Cargo) to have `examples` folder, so you can do `cargo run --example echo`.
> So it felt natural to have frontend directory nearby.

Perfect then :)

> > Why is mediasoup rust version 0.5.0?
> 
> This is less similar to TypeScript library and more similar to mediasoup-client case.
> Rust version will have different things that will be considered a breaking change (for instance #503 broke dump in Rust because they are heavily typed there) and I don't think versioning will line-up perfectly.
> And there will probably be more things that will break until it reaches the level of API stability of mediasoup itself.

Ok, let's do this way for now and let's see in the future.

NOTE: I still have to review more things and won't be able to finish this week. Will do next one, promise.

nazar-pc committed:
Merge remote-tracking branch 'upstream/v3' into rust

nazar-pc committed:
Rename workflow file similarly to upstream, remove custom branch, add badge to readme

nazar-pc committed:
Improve error handling for cases when file descriptors for channel and/or payload channel is already closed

nazar-pc committed:
Cleanup extra nested modules in integration tests

nazar-pc committed:
Merge remote-tracking branch 'upstream/v3' into rust-updates

nazar-pc committed:
Bump version

nazar-pc committed:
Make CI run on all commits and PRs

nazar-pc committed:
Replace unwrap with error handling

nazar-pc commented:
Tests are still a bit flaky at times, but I'll probably be ironing them out in next version instead of introducing more changes to this PR.

ibc commented on worker/src/RTC/RtpDictionaries/RtpEncodingParameters.cpp:
```diff
@@ -151,9 +151,6 @@ namespace RTC
 		if (!this->scalabilityMode.empty())
 		{
 			jsonObject["scalabilityMode"] = this->scalabilityMode;
-			jsonObject["spatialLayers"]   = this->spatialLayers;
-			jsonObject["temporalLayers"]  = this->temporalLayers;
-			jsonObject["ksvc"]            = this->ksvc;
```
Can we please keep those? We don't need to remove them from `dump()` method whose output is not even documented (private API just for debugging). We really need them for debugging.

ibc replied:
I'm gonna do it

ibc replied:
done here: https://github.com/versatica/mediasoup/pull/518/commits/b2047f73512197086ec1129098c4be8e6fdf8e2b

ibc commented on worker/src/RTC/SctpAssociation.cpp:
```diff
@@ -356,15 +356,6 @@ namespace RTC
 		// Add maxMessageSize.
 		jsonObject["maxMessageSize"] = this->maxSctpMessageSize;
-
-		// Add sendBufferSize.
-		jsonObject["sendBufferSize"] = this->sctpSendBufferSize;
-
-		// Add sctpBufferedAmountLowThreshold.
-		jsonObject["sctpBufferedAmount"] = this->sctpBufferedAmount;
-
-		// Add isDataChannel.
-		jsonObject["isDataChannel"] = this->isDataChannel;
```
Same. Can we please keep those? We don't need to remove them from `dump()` method whose output is not even documented (private API just for debugging). We really need them for debugging.

ibc replied:
I'm gonna do it

ibc replied:
done here: https://github.com/versatica/mediasoup/pull/518/commits/b2047f73512197086ec1129098c4be8e6fdf8e2b

ibc reviewed (changes requested):
Please address just these two comments and let's merge.

ibc committed:
Revert properties removal in RtpEncodingParameters::FillJson() and SctpAssociation::FillJson()

ibc commented:
waiting for tests to pass in CI
